### PR TITLE
render duration of intro & outro ranges on overview waveforms

### PR DESCRIPTION
**Pull request by Be-ing**
MarkRange elements in skins only have their durations rendered if they specify a DurationTextColor. This avoids rendering the durations of loops on the overview waveform.

image

durations scale with rate slider:
image

I played a 3 hour set with this last night and it is extremely helpful for me. Having Mixxx calculate these times for me allows me to be more intentional about how I align tracks. I used to do this in headphones continually. With this I don't have to keep the times in my head, I don't have to do any math, and I don't have to repeat that the next time I play a track.

I understand this could be distracting or clutter for others, so I'll add an option to disable this unless there are objections to adding an option.

### CODE REVIEW

File: src/widget/woverview.cpp [hunk 1]
```diff
@@ -565,6 +533,65 @@ void WOverview::paintEvent(QPaintEvent * /*unused*/) {
...
+                    painter.drawRect(QRectF(QPointF(-2.0, startPosition),
+                                            QPointF(width() + 1.0, endPosition)));
+                }
+
+                // draw duration of range
+                if (markRange.showDuration()) {
+                    // TODO: replace with rate_ratio in PR #1765
+                    double rateRatio = 1.0 + m_pRateDirControl->get() * m_pRateRangeControl->get() * m_pRateSliderControl->get();
+                    QString duration = mixxx::Duration::formatTime((endValue - startValue)
+                            / m_trackSampleRateControl->get() / 2 / rateRatio);
```
Comment (uklotzde) [inline src/widget/woverview.cpp]: It might be worth to define a constant for the number of channels to not forget about it once the engine is rewritten.

**Comment by uklotzde** [comment]
Very helpful feature :+1:

Instead of only enabling/disabling the display we could think about specifying a minimum duration, e.g. with a configurable range between 0 to 30 seconds and a default value of 3 or 5 seconds. This value may vary depending on your music and mixing style. Not sure if this is more confusing than helpful?

Now I also noticed that I need markers on the overview waveforms for a quick peek how long a phrase might be :see_no_evil:

**Comment by Be-ing** [comment]
> Not sure if this is more confusing than helpful?

I think that would be more confusing than helpful.

**Comment by daschuer** [comment]
How about this appearance: 
grafik

We can move the time label conditionally into the region if enough space is available. 

I don't think we need a preferences option for it.

**Comment by Be-ing** [comment]
Why do you want the text moved outside the range? And why move it to the top?

If we can find a solution that everyone is satisfied with that does not require a preference option that would be ideal.

**Comment by daschuer** [comment]
It look IMHO ugly if the time labele escapes the boundaries of the highlighted region. 
Like proposed it labels the triangle "00:15 ◣"  and "◢ 00:23" which keeps all the eye candy in one place.

**Comment by Be-ing** [comment]
To clarify, are you proposing putting the text outside the range on the side closer to the middle of the track? So for the intro it would be drawn after the range and for the outro it would be drawn before the range?

**Comment by uklotzde** [comment]
This is exactly how I understood Daniel's proposal. The labels would be displayed outside of the (probably short region) within the main track range, next to their nearest marker.

For short intro/outro ranges the labels might not fit into their sections and would otherwise be displayed across the boundaries with varying backgrounds.

There is no ideal solution. With Daniel's proposal the appearance is slightly more consistent.

**Comment by uklotzde** [comment]
A typical example of a short intro:
Screenshot from 2019-05-01 20-41-35

Reusing the translucent background from highlighted region for the label might be an option to improve readability. The waveform will still shine through.

**Comment by daschuer** [comment]
> To clarify, are you proposing putting the text outside the range on the side closer to the middle of the track? So for the intro it would be drawn after the range and for the outro it would be drawn before the range?

Yes, as an idea for discussion. The part I like most is the idea to have the texts and the triangle close to each other and not draw across the region borders.

**Comment by Be-ing** [comment]
How is this? Do we still need to add a preference option?
image

**Comment by daschuer** [comment]
Ok, thank you. 
Is the ◣ under the blue overlay? It should have the same color than the time text.

**Comment by Be-ing** [comment]
Good catch, fixed:
image

**Comment by daschuer** [comment]
LGTM Thanks, waiting for CI.

**Comment by Be-ing** [comment]
CI passed :+1:

**Comment by Be-ing** [comment]
If there is still a demand for adding an option to disable this, that can still be added during the 2.3 beta period. Hopefully this looks good enough that it isn't bothersome enough to anyone that an option is required.

**Pull request merged**